api/src/main/java/com/epam/pipeline/manager/datastorage/lifecycle/DataStorageLifecycleRestoreManager.java

```diff
@@ -0,0 +1,254 @@
+        final List<StorageRestoreAction> relatedActions = filterRestoreStorageActions(
+                storage,
+                StorageRestoreActionSearchFilter.builder()
+                        .datastorageId(storage.getId())
+                        .path(path)
+                        .searchType(StorageRestoreActionSearchFilter.SearchType.SEARCH_PARENT)
+                        .statuses(StorageRestoreStatus.ACTIVE_STATUSES)
+                        .build()
+        );
+        return relatedActions.stream()
+                .filter(a -> a.getStatus() == StorageRestoreStatus.SUCCEEDED
+                        && a.getRestoredTill().isAfter(DateUtils.nowUTC())
+                ).findFirst().orElse(relatedActions.stream().findFirst().orElse(null));
+    }
+
+    public List<StorageRestoreAction> loadEffectiveRestoreStorageActionHierarchy(final AbstractDataStorage storage,
+                                                                                 final StorageRestorePath path,
+                                                                                 final Boolean recursive) {
+        final StorageRestoreAction root = loadEffectiveRestoreStorageAction(storage, path);
+        final boolean rootActionIsActive = root != null && (isActionStillActive(root));
+        if (path.getType() == StorageRestorePathType.FILE) {
+            return rootActionIsActive ? Collections.singletonList(root) : Collections.emptyList();
+        }
+        final List<StorageRestoreAction> relatedActions = filterRestoreStorageActions(
+                storage,
+                StorageRestoreActionSearchFilter.builder()
+                        .datastorageId(storage.getId())
+                        .path(path)
+                        .isLatest(true)
+                        .statuses(StorageRestoreStatus.ACTIVE_STATUSES)
+                        .searchType(recursive
+                                ? StorageRestoreActionSearchFilter.SearchType.SEARCH_CHILD_RECURSIVELY
+                                : StorageRestoreActionSearchFilter.SearchType.SEARCH_CHILD
+                        ).build())
+                .stream()
+                // check if root action override child
+                .filter(childAction -> !rootActionIsActive || childAction.getStarted().isAfter(root.getStarted()))
+                // check if child action still active
+                .filter(DataStorageLifecycleRestoreManager::isActionStillActive).collect(Collectors.toList());
+
+        return ListUtils.union(
+                rootActionIsActive ? Collections.singletonList(root) : Collections.emptyList(),
+                relatedActions
+        );
+    }
+
+    @Transactional
+    public void deleteRestoreActions(final Long datastorageId) {
+        dataStoragePathRestoreActionRepository.deleteByDatastorageId(datastorageId);
+        dataStoragePathRestoreActionRepository.flush();
```

Comment:
Do we really need to manually flush?

Reply:
Yes, it's because we need to force hibernate to propagate changes that were done to DB, because this method is used by StorageManager when it deletes some storage and StorageManager uses jdbc to connect to db, so as I see there are two ways:
 - `flush` hibernate changes to DB, in this case this changes will still under common transaction 
 - provide for this method `propogation = REQUIRED_NEW` - but here we will commit all changes to db
 
 Maybe I miss something and there is another convenient way to do that